File: sdk/js/test/response-contract.test.ts
// Copyright 2026 STARGA, Inc.
import { it } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { MindMemClient } from "../src/client.js";

it("preserves actual REST envelopes, ranked hits and evidence", async () => {
  // Paths are resolved from sdk/js, the documented npm test working directory.
  const fixtures = Object.fromEntries(
    ["recall", "block", "health", "contradictions", "scan"].map(name => [
      name, JSON.parse(readFileSync(`../spec/fixtures/${name}.json`, "utf8")),
    ]),
  );
  const original = globalThis.fetch;
  const requested: Record<string, unknown>[] = [];
  globalThis.fetch = async (url, options) => {
    const path = new URL(String(url)).pathname;
    const name = path.startsWith("/v1/block/") ? "block" : path.split("/").at(-1)!;
    if (options?.body) requested.push(JSON.parse(String(options.body)));
    return new Response(JSON.stringify(fixtures[name]), { status: 200 });
  };
  try {
    const client = new MindMemClient("http://fixture.invalid");
    const recall = await client.recall("orchid", { scoringInstant: "2026-09-14" });
    assert.equal(recall.count, 1);
    assert.equal(recall.results[0]?._id, "D-20260914-001");
    assert.equal(recall.results[0]?.excerpt, "Orchid is the SDK contract sentinel.");
    assert.deepEqual(recall.attestation, fixtures["recall"].attestation);
    assert.equal(requested[0]?.["scoring_instant"], "2026-09-14");
    const block = await client.getBlock("D-20260914-001");
    assert.equal(block.found, true);
    assert.equal(block.block.Statement, "Orchid is the SDK contract sentinel.");
    const health = await client.health();
    assert.equal(health.workspace_exists, true);
    assert.ok(health.api_version);
    const contradictions = await client.listContradictions();
    assert.equal(contradictions.contradictions, 0);
    assert.equal(contradictions.status, "clean");
    const scan = await client.scan();
    assert.deepEqual(scan.checks["decisions"], { total: 1, active: 1 });
  } finally {
    globalThis.fetch = original;
  }
});


it("represents anonymous health without private workspace fields", async () => {
  const fixture = JSON.parse(readFileSync("../spec/fixtures/health_public.json", "utf8"));
  const original = globalThis.fetch;
  globalThis.fetch = async () => new Response(JSON.stringify(fixture), { status: 200 });
  try {
    const health = await new MindMemClient("http://fixture.invalid").health();
    assert.ok(health.api_version);
    assert.equal(health.workspace, undefined);
    assert.equal(health.workspace_exists, undefined);
  } finally {
    globalThis.fetch = original;
  }
});

File: sdk/go/response_contract_test.go
// Copyright 2026 STARGA, Inc.
package mindmem_test

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"testing"

	mindmem "github.com/star-ga/mind-mem/sdk/go/v5"
)

func contractServer(t *testing.T, name string) (*mindmem.Client, func()) {
	t.Helper()
	raw, err := os.ReadFile("testdata/contract/" + name + ".json")
	if err != nil {
		t.Fatal(err)
	}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write(raw)
	}))
	return mindmem.NewClient(srv.URL), srv.Close
}

func TestActualServerRecallContract(t *testing.T) {
	c, close := contractServer(t, "recall")
	defer close()
	result, err := c.Recall(context.Background(), "orchid", mindmem.RecallOptions{ScoringInstant: "2026-09-14"})
	if err != nil {
		t.Fatal(err)
	}
	if result.Count != 1 || len(result.Results) != 1 || result.Results[0].ID != "D-20260914-001" || result.Results[0].Excerpt != "Orchid is the SDK contract sentinel." {
		t.Fatalf("lost served result: %+v", result)
	}
	if result.SchemaVersion != "1.0" || result.ScoringInstant != "2026-09-14" {
		t.Fatalf("lost response identity: %+v", result)
	}
	var attestation map[string]json.RawMessage
	if err := json.Unmarshal(result.Attestation, &attestation); err != nil {
		t.Fatal(err)
	}
	if string(attestation["schema"]) != `"RECALL_ATTEST_v2"` {
		t.Fatal("lost attestation")
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(result.Raw, &raw); err != nil {
		t.Fatal(err)
	}
	if string(raw["attestation"]) != string(result.Attestation) {
		t.Fatal("raw evidence changed")
	}
}

func TestActualServerBlockContract(t *testing.T) {
	c, close := contractServer(t, "block")
	defer close()
	result, err := c.GetBlock(context.Background(), "D-20260914-001")
	if err != nil {
		t.Fatal(err)
	}
	if !result.Found || result.Block.ID != result.BlockID || result.Block.Statement != "Orchid is the SDK contract sentinel." || result.Block.Status != "active" {
		t.Fatalf("lost governed fields: %+v", result)
	}
	if len(result.Raw) == 0 {
		t.Fatal("raw optional fields were discarded")
	}
}

func TestActualServerDiagnosticContracts(t *testing.T) {
	c, close := contractServer(t, "contradictions")
	contradictions, err := c.ListContradictions(context.Background())
	close()
	if err != nil {
		t.Fatal(err)
	}
	if contradictions.Status != "clean" || contradictions.Contradictions != 0 {
		t.Fatalf("wrong count envelope: %+v", contradictions)
	}
	c, close = contractServer(t, "health")
	health, err := c.Health(context.Background())
	close()
	if err != nil {
		t.Fatal(err)
	}
	if health.WorkspaceExists == nil || !*health.WorkspaceExists || health.Workspace == nil || health.APIVersion == "" || health.WorkspaceSchemaVersion == "" {
		t.Fatalf("lost health fields: %+v", health)
	}
	c, close = contractServer(t, "scan")
	scan, err := c.Scan(context.Background())
	close()
	if err != nil {
		t.Fatal(err)
	}
	var counts map[string]int
	if err := json.Unmarshal(scan.Checks["decisions"], &counts); err != nil {
		t.Fatal(err)
	}
	if counts["active"] != 1 || scan.Backend != "markdown" {
		t.Fatalf("lost scan checks: %+v", scan)
	}
}

func TestPublicHealthPreservesAbsence(t *testing.T) {
	c, close := contractServer(t, "health_public")
	defer close()
	health, err := c.Health(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if health.APIVersion == "" || health.WorkspaceExists != nil || health.Workspace != nil {
		t.Fatalf("public health must omit private workspace fields: %+v", health)
	}
}
